src/librbd/image/CloneRequest.cc

```diff
@@ -0,0 +1,602 @@
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_close() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  assert(m_imctx != nullptr);
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_async_context_callback(
+    *m_imctx, create_context_callback<
+      klass, &klass::handle_close, false>(this));
+  m_imctx->state->close(ctx);
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_close(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't close image: " << cpp_strerror(*r) << dendl;
+    return m_on_finish;
+  }
+
+  switch_thread_context();
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::switch_thread_context() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = CloneRequest<I>;
+
+  Context *ctx = create_context_callback<klass, &klass::handle_switch_thread_context>(this);
+  m_op_work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void CloneRequest<I>::handle_switch_thread_context(int r) {
+  ldout(m_cct, 20) << ": r=" << r << dendl;
+
+  m_imctx->destroy();
+  m_imctx = nullptr;
+
+  if (m_r_saved == 0) {
+    complete(0);
+  } else {
+    send_remove();
+  }
+  return;
+}
+
+template <typename I>
+void CloneRequest<I>::send_remove_child() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::remove_child(&op, m_pspec, m_id);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_callback<klass, &klass::handle_remove_child>(this);
+  int r = m_p_imctx->md_ctx.aio_operate(RBD_CHILDREN, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_remove_child(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r < 0) {
+     lderr(m_cct) << "Error removing failed clone from list of children: "
+		  << cpp_strerror(*r) << dendl;
+  }
+
+  send_close();
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_remove() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_remove>(this);
+
+  librbd::image::RemoveRequest<> *req = librbd::image::RemoveRequest<>::create(
+   m_ioctx, m_name, m_id, false, m_no_op, m_op_work_queue, ctx);
+  req->send();
+}
+
+template <typename I>
+void CloneRequest<I>::handle_remove(int r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "Error removing failed clone: "
+		 << cpp_strerror(r) << dendl;
+  }
+  return complete(r);
+}
+
+template <typename I>
+void CloneRequest<I>::complete(int r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << r << dendl;
+
+  if (r == 0) {
+    ldout(m_cct, 20) << "done." << dendl;
+  }
+
+  m_on_finish->complete(r);
+  delete this;
```

Comment:
Thanx I use the ```Context *handle_xxx(int *)``` style in the whole state machine, same with other state machine in librbd/image/